### PR TITLE
Ndev 3100 use evm steps for tracing

### PR DESCRIPTION
**Pull request by ancientmage**

### CODE REVIEW

File: evm_loader/program/src/types/execution_map.rs
Comment (evgenyzdanovich) [inline evm_loader/program/src/types/execution_map.rs]: self.steps.iter().any(|&s| s.is_reset)

File: evm_loader/lib/src/commands/emulate.rs
Comment (evgenyzdanovich) [inline evm_loader/lib/src/commands/emulate.rs]: tiny: how about `solana_accounts` instead of `solanas`?

File: evm_loader/lib/src/commands/emulate.rs
Comment (evgenyzdanovich) [inline evm_loader/lib/src/commands/emulate.rs:409]: optional suggestion: how about reusing `emulate_trx_single_step` inside the `emulate_trx_multiple_steps`, is it possible?
You likely can move the call of `calculate_response` from the `emulate_trx_single_step` into the L273 and L279 

If it's too clunky/difficult to do, drop this suggestion.

**Review by evgenyzdanovich — DISMISSED** [review]
Soft Approval from my side, only minor comments. 
Please wait for Semen's approval.

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11255533628/artifacts/2035057393)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   67443 |      0     |      0.165 |    0.000578652 |         0     |
| getAccountInfo                    |   57251 |      0     |      0.098 |    0.00096573  |         0     |
| getBlocks                         |   26558 |      0     |      0.049 |    0.00044721  |         0     |
| getBlock                          |   15673 |      0     |      0.04  |    0.000965163 |         0.001 |
| getMultipleAccounts               |   13471 |      0     |      0.112 |    0.0013509   |         0.001 |
| getLatestBlockhash                |   10995 |      0     |      0.104 |    0.00107076  |         0     |
| getTransaction                    |    7293 |      0     |      0.186 |    0.0076653   |         0.003 |
| getBlockTime                      |    6021 |      0     |      0.05  |    0.000325527 |         0     |
| sendTransaction                   |    5022 |      0     |      0.096 |    0.00564417  |         0.003 |
| getBalance                        |    1825 |      0     |      0.037 |    0.000955616 |         0     |
| getSignatureStatuses              |    1133 |      0     |      0.065 |    0.00078376  |         0     |
| getVersion                        |     599 |      0     |      0.025 |    0.000949917 |         0     |
| isBlockhashValid                  |     548 |      0     |      0.013 |    0.000301095 |         0     |
| getSignaturesForAddress           |      63 |      0     |      0.004 |    0.00101587  |         0.001 |
| simulateTransaction               |      38 |      0     |      0.001 |    0.000315789 |         0     |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.000625    |         0.001 |
| getTokenAccountBalance            |       7 |      0     |      0.003 |    0.000714286 |         0     |
| getHealth                         |       3 |      0     |      0.004 |    0.00133333  |         0     |
| getFirstAvailableBlock            |       2 |      0     |      0     |    0           |         0     |
| getMinimumBalanceForRentExemption |       1 |      0     |      0     |    0           |         0     |
| getClusterNodes                   |       1 |      0.001 |      0.001 |    0.001       |         0.001 |

</details>

File: evm_loader/lib/src/commands/emulate.rs
Comment (ancientmage) [reply evm_loader/lib/src/commands/emulate.rs:409]: If I try to do so than 
`if exit_status == ExitStatus::StepLimit` must be outside of emulate_trx_single_step, otherwise it will give incorrect result for multiple steps where on almost every instruction you will receive step_limit.
Also, creation of backend & evm must be outside too because of the logic of a reinitialization (there is 

So, what's have left? Only call of evm::execute. I don't think it's a good idea

**Comment by evgenyzdanovich** [comment]
Looks good for me, let's wait for Semen's opinion

**Review by evgenyzdanovich — DISMISSED** [review]

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11353689599/artifacts/2060258798)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   59825 |      0     |      0.14  |    0.000724914 |         0     |
| getAccountInfo                    |   57186 |      0     |      0.097 |    0.00123425  |         0     |
| getBlocks                         |   23471 |      0     |      0.057 |    0.000572153 |         0     |
| getBlock                          |   15135 |      0     |      0.044 |    0.00121513  |         0.001 |
| getMultipleAccounts               |   13443 |      0     |      0.139 |    0.00169412  |         0.001 |
| getLatestBlockhash                |   10730 |      0     |      0.106 |    0.00117819  |         0     |
| getTransaction                    |    7315 |      0     |      0.113 |    0.00908339  |         0.004 |
| getBlockTime                      |    6114 |      0     |      0.084 |    0.00041086  |         0     |
| sendTransaction                   |    5027 |      0     |      0.399 |    0.00653551  |         0.003 |
| getBalance                        |    1821 |      0     |      0.033 |    0.00121087  |         0     |
| getSignatureStatuses              |    1126 |      0     |      0.085 |    0.00101066  |         0     |
| getVersion                        |     587 |      0     |      0.022 |    0.000868825 |         0     |
| isBlockhashValid                  |     545 |      0     |      0.019 |    0.000502752 |         0     |
| getSignaturesForAddress           |      62 |      0     |      0.003 |    0.00117742  |         0.001 |
| simulateTransaction               |      38 |      0     |      0.001 |    0.000263158 |         0     |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.0005625   |         0.001 |
| getHealth                         |       3 |      0     |      0.001 |    0.000333333 |         0     |
| getTokenAccountBalance            |       3 |      0     |      0.001 |    0.000333333 |         0     |
| getFirstAvailableBlock            |       2 |      0     |      0     |    0           |         0     |
| getMinimumBalanceForRentExemption |       1 |      0     |      0     |    0           |         0     |
| getClusterNodes                   |       1 |      0.002 |      0.002 |    0.002       |         0.002 |

</details>

File: evm_loader/lib/src/commands/emulate.rs
Comment (s-medvedev) [inline evm_loader/lib/src/commands/emulate.rs]: I don't like the `flag` arguments which change the function internal logic. It's hard to keep in mind all possible combinations. I think in this case, the code will be clearer if you simply call the `increment_nonce` and `transfer_gas_limit` functions in the initialization places.

Comment (ancientmage) [reply evm_loader/lib/src/commands/emulate.rs]: fixed

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11403894746/artifacts/2074176954)

**Review by s-medvedev — DISMISSED** [review]

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11487533272/artifacts/2095839447)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   58146 |      0     |      0.093 |    0.000766932 |         0     |
| getAccountInfo                    |   53995 |      0     |      0.118 |    0.00134631  |         0     |
| getBlocks                         |   23404 |      0     |      0.05  |    0.000569347 |         0     |
| getBlock                          |   15166 |      0     |      0.164 |    0.00132883  |         0.001 |
| getMultipleAccounts               |   13080 |      0     |      0.069 |    0.00176468  |         0.001 |
| getLatestBlockhash                |   10502 |      0     |      0.078 |    0.00138231  |         0     |
| getTransaction                    |    7303 |      0     |      0.135 |    0.0100874   |         0.004 |
| getBlockTime                      |    5879 |      0     |      0.061 |    0.000451948 |         0     |
| sendTransaction                   |    5003 |      0     |      0.095 |    0.00787587  |         0.004 |
| getBalance                        |    1745 |      0     |      0.075 |    0.00135186  |         0     |
| getSignatureStatuses              |    1106 |      0     |      0.018 |    0.000898734 |         0     |
| getVersion                        |     577 |      0     |      0.023 |    0.000958406 |         0     |
| isBlockhashValid                  |     535 |      0     |      0.027 |    0.000403738 |         0     |
| simulateTransaction               |     491 |      0     |      0.004 |    0.00104073  |         0.001 |
| getSignaturesForAddress           |      58 |      0     |      0.003 |    0.00127586  |         0.001 |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.0006875   |         0.001 |
| getTokenAccountBalance            |       4 |      0     |      0.001 |    0.00025     |         0     |
| getHealth                         |       3 |      0     |      0.001 |    0.000666667 |         0.001 |
| getFirstAvailableBlock            |       2 |      0     |      0     |    0           |         0     |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |         0.001 |
| getClusterNodes                   |       1 |      0.005 |      0.005 |    0.005       |         0.005 |

</details>

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11575760245/artifacts/2118176025)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   60666 |      0     |      0.169 |    0.000993489 |        0.001  |
| getAccountInfo                    |   55169 |      0     |      0.222 |    0.00204363  |        0.001  |
| getBlocks                         |   25195 |      0     |      0.202 |    0.000705616 |        0.001  |
| getBlock                          |   16447 |      0     |      0.154 |    0.0018213   |        0.002  |
| getMultipleAccounts               |   13733 |      0     |      0.122 |    0.00229753  |        0.001  |
| getTransaction                    |   12649 |      0     |      0.228 |    0.0118015   |        0.004  |
| getLatestBlockhash                |   11071 |      0     |      0.15  |    0.00172026  |        0.001  |
| getBlockTime                      |    6039 |      0     |      0.142 |    0.000782911 |        0      |
| sendTransaction                   |    5167 |      0     |      0.198 |    0.0081167   |        0.004  |
| getBalance                        |    1773 |      0     |      0.056 |    0.00189397  |        0.001  |
| getSignatureStatuses              |    1134 |      0     |      0.205 |    0.00225573  |        0.001  |
| getVersion                        |     600 |      0     |      0.042 |    0.00225167  |        0.001  |
| isBlockhashValid                  |     549 |      0     |      0.024 |    0.000531876 |        0      |
| simulateTransaction               |     491 |      0     |      0.007 |    0.00180652  |        0.002  |
| getSignaturesForAddress           |      59 |      0     |      0.003 |    0.00161017  |        0.002  |
| requestAirdrop                    |      32 |      0     |      0.002 |    0.0009375   |        0.001  |
| getTokenAccountBalance            |       8 |      0     |      0.016 |    0.002375    |        0.0005 |
| getHealth                         |       3 |      0     |      0.001 |    0.000666667 |        0.001  |
| getFirstAvailableBlock            |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getMinimumBalanceForRentExemption |       1 |      0.003 |      0.003 |    0.003       |        0.003  |
| getClusterNodes                   |       1 |      0     |      0     |    0           |        0      |

</details>

**Review by s-medvedev — APPROVED** [review]

**Pull request merged**
